homeassistant/components/github/config_flow.py

```diff
@@ -0,0 +1,195 @@
+    CONF_ACCESS_TOKEN,
+    CONF_REPOSITORIES,
+    DEFAULT_REPOSITORIES,
+    DOMAIN,
+    LOGGER,
+)
+
+
+async def starred_repositories(hass: HomeAssistant, access_token: str) -> list[str]:
+    """Return a list of repositories that the user has starred."""
+    client = GitHubAPI(token=access_token, session=async_get_clientsession(hass))
+
+    async def _get_starred():
+        response = await client.user.starred(**{"params": {"per_page": 100}})
+        if not response.is_last_page:
+            results = await asyncio.gather(
+                *(
+                    client.user.starred(
+                        **{"params": {"per_page": 100, "page": page_number}},
+                    )
+                    for page_number in range(
+                        response.next_page_number, response.last_page_number + 1
+                    )
+                )
+            )
+            for result in results:
+                response.data.extend(result.data)
+
+        return response.data
+
+    try:
+        if len(result := await _get_starred()) == 0:
+            return DEFAULT_REPOSITORIES
+
+        return sorted((repo.full_name for repo in result), key=str.casefold)
+    except GitHubException:
+        pass
+
+    return DEFAULT_REPOSITORIES
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for GitHub."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize."""
+        self._device: GitHubDeviceAPI | None = None
+        self._login: GitHubLoginOauthModel | None = None
+        self._login_device: GitHubLoginDeviceModel | None = None
+
+    async def async_step_user(
+        self,
+        user_input: dict[str, Any] | None = None,
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if self._async_current_entries():
+            return self.async_abort(reason="already_configured")
+
+        return await self.async_step_device(user_input)
+
+    async def async_step_device(self, user_input: dict[str, Any] | None = None):
+        """Handle device steps."""
+
+        async def _wait_for_login():
+            response = await self._device.activation(
+                device_code=self._login_device.device_code
+            )
+            self._login = response.data
+            await self.hass.config_entries.flow.async_configure(flow_id=self.flow_id)
+
+        if self._login:
+            return self.async_show_progress_done(next_step_id="repositories")
+
+        if not self._device:
+            self._device = GitHubDeviceAPI(
+                client_id=CLIENT_ID,
+                session=async_get_clientsession(self.hass),
+                **{"client_name": SERVER_SOFTWARE},
+            )
+        try:
+            response = await self._device.register()
+            self._login_device = response.data
+        except GitHubException as exception:
+            LOGGER.exception(exception)
+            return self.async_abort(reason="could_not_register")
```

Comment:
Do we need to finish the progress before aborting?

https://github.com/home-assistant/core/blob/64bcd6097457b0c56528425a6a6ce00a2bce791c/homeassistant/data_entry_flow.py#L268

Reply:
Not sure, I'll test that

Reply:
Not needed.